# Train word2vec locally

This allows a smart initialization of our neural net's word embeddings.
It seems that initializing the embeddings by training them locally, as opposed to using pre-trained word2vec embeddings (available online) can lead to better performance.

In [59]:
import os
import sys
print(sys.executable)

/Users/andrei/anaconda3/envs/cil/bin/python


In [60]:
from gensim.models.word2vec import Word2Vec

import logging
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)

In [79]:
TRAIN = os.path.join('..', 'train')
TEST = os.path.join('..', 'test')
POS_TWEET_FILE = os.path.join(TRAIN, 'train_pos_full.txt')
NEG_TWEET_FILE = os.path.join(TRAIN, 'train_pos_full.txt')
TEST_TWEET_FILE = os.path.join(TEST, 'test_data.txt')
EMBEDDING_SIZE = 400

In [80]:
def read_tweets(fname):
    """Read the tweets in the given file.
    
    Returns a 2d array where every row is a tweet, split into words.
    """
    with open(fname, 'r') as f:
        return [l.split() for l in f.readlines()]

In [63]:
pos_tweets = read_tweets(POS_TWEET_FILE)

In [64]:
neg_tweets = read_tweets(NEG_TWEET_FILE)

In [65]:
test_tweets = read_tweets(TEST_TWEET_FILE)

In [66]:
sentences = pos_tweets + neg_tweets + test_tweets
print(len(sentences))

2510000


In [81]:
# Download this for testing: https://github.com/arfon/word2vec/blob/master/questions-words.txt
# Highly recommended!

question_file = "questions-words.txt"

def eval_embeddings(model):
    accuracy_results = model.accuracy(question_file)
    summary = accuracy_results[-1]
    assert summary['section'] == 'total'
    incorrect = summary['incorrect']
    correct = summary['correct']

    incorrect_n = len(incorrect)
    correct_n = len(correct)

    acc = correct_n / incorrect_n
    return acc, correct_n, incorrect_n

In [82]:
WORKERS = 8
# Note: Moises's team uses size=200 as of June 13.
# See: https://groups.google.com/forum/#!msg/gensim/ggCHGncd5rU/Z_pQDD69AAAJ
# for some parameter hints.
model = Word2Vec(sentences, size=EMBEDDING_SIZE, window=10, min_count=5, workers=WORKERS)# , alpha=0.05, cbow_mean=1)

In [83]:
# Should be queen
model.most_similar(positive=['woman', 'king'], negative=['man'])

[('queen', 0.48789501190185547),
 ('madonna', 0.3770555853843689),
 ('hawk', 0.3623706102371216),
 ('alpha', 0.353135883808136),
 ('kenny', 0.3516470193862915),
 ('sinner', 0.33579689264297485),
 ('poet', 0.33149391412734985),
 ('citizen', 0.3234192728996277),
 ('lawrence', 0.32013171911239624),
 ('chief', 0.3183271884918213)]

In [84]:
# Should be germany
model.most_similar(positive=['france', 'berlin'], negative=['paris'])

[('croatia', 0.62614506483078),
 ('germany', 0.5872930288314819),
 ('spain', 0.586173415184021),
 ('switzerland', 0.5797890424728394),
 ('russia', 0.579714298248291),
 ('finland', 0.5775009393692017),
 ('belgium', 0.5770790576934814),
 ('norway', 0.5755430459976196),
 ('sweden', 0.5712461471557617),
 ('portugal', 0.5701516270637512)]

In [85]:
model.doesnt_match("breakfast cereal dinner lunch".split())

'cereal'

In [86]:
model.estimate_memory()

{'syn0': 111705600,
 'syn1neg': 111705600,
 'total': 258319200,
 'vocab': 34908000}

In [87]:
# A few more sanity checks
print(model.similarity('woman', 'man'))
print(model.similarity('woman', 'coffee'))
print(model.similarity('woman', 'penis'))
print(model.similarity('woman', 'football'))

0.498196601473
0.103336089782
0.225357085838
0.0123211443995


In [88]:
print(model.similarity('car','man'))
print(model.similarity('car','truck'))

0.186135759668
0.665023399397


In [89]:
acc, correct_n, incorrect_n = eval_embeddings(model)
print("{0:5.3f} accuracy; Analogies: {1} correct, {2} incorrect".format(
    acc, correct_n, incorrect_n))

0.341 accuracy; Analogies: 2340 correct, 6870 incorrect


### Accuracies (full Twitter data)
 * Vanilla (size=225, window=5, min_count=5): 0.319 accuracy; Analogies: 2233 correct, 7004 incorrect
 * (size=300, window=5,  min_count=5): 0.337 accuracy; Analogies: 2329 correct, 6908 incorrect
 * (size=500, window=5,  min_count=5): 0.330 accuracy; Analogies: 2292 correct, 6945 incorrect
 * (size=300, window=10, min_count=5): 0.346 accuracy; Analogies: 2374 correct, 6863 incorrect
 * (size=300, window=15, min_count=5): 0.342 accuracy; Analogies: 2356 correct, 6881 incorrect
 * (size=400, window=10, min_count=5): 0.341 accuracy; Analogies: 2340 correct, 6870 incorrect


In [90]:
print("Embedding dimensionality: {0}".format(EMBEDDING_SIZE))

Embedding dimensionality: 400


In [91]:
fname = "./word2vec-local-gensim-{0}.bin".format(EMBEDDING_SIZE)
model.save(fname)

### Some experimentation

In [26]:
emb_sizes = [100, 150, 175, 200, 225, 250, 275, 300, 325, 350, 375, 400]

for emb_size in emb_sizes:
    print("Computing embeddings of size {0}...".format(emb_size))
    model = Word2Vec(sentences, size=emb_size, window=10, min_count=5, workers=WORKERS)
    print("Evaluating embeddings of size {0}...".format(emb_size))
    acc, correct_n, incorrect_n = eval_embeddings(model)
    print("Size {3}: {0:5.3f} accuracy; Analogies: {1} correct, {2} incorrect".format(
        acc, correct_n, incorrect_n, emb_size))

Computing embeddings of size 100...
Evaluating embeddings of size 100...
Size 100: 0.274 accuracy; Analogies: 1983 correct, 7227 incorrect
Computing embeddings of size 150...
Evaluating embeddings of size 150...
Size 150: 0.317 accuracy; Analogies: 2216 correct, 6994 incorrect
Computing embeddings of size 175...
Evaluating embeddings of size 175...
Size 175: 0.323 accuracy; Analogies: 2246 correct, 6964 incorrect
Computing embeddings of size 200...
Evaluating embeddings of size 200...
Size 200: 0.337 accuracy; Analogies: 2323 correct, 6887 incorrect
Computing embeddings of size 225...
Evaluating embeddings of size 225...
Size 225: 0.336 accuracy; Analogies: 2316 correct, 6894 incorrect
Computing embeddings of size 250...
Evaluating embeddings of size 250...
Size 250: 0.337 accuracy; Analogies: 2319 correct, 6891 incorrect
Computing embeddings of size 275...
Evaluating embeddings of size 275...
Size 275: 0.346 accuracy; Analogies: 2369 correct, 6841 incorrect
Computing embeddings of siz